### ISSUE TITLE
Web container crashes when task does not have a file or config parameter

### ISSUE THREAD

**Issue by gijsboer** [concourse/concourse#1739]
# Bug Report

I accidentally created a pipeline that didn't have a file nor a config parameter for a task, and the web container crashed because of it. Ended up debugging it for hours to find out what was actually wrong. Whenever it tries to run the task, the entire web container crashes, it then restarts itself and then crashes again, stuck in an endless crashing loop.

As far as I could see, there was no way to remove the pipeline causing the problem, preventing the web container from crashing again. So I needed to remove all the data and fill in the pipeline again. While that in itself is not a big problem, but if people want to keep their build logs, then I can't see a way of doing that.

Actual result:
```
concourse-web_1     | panic: runtime error: invalid memory address or nil pointer dereference
concourse-web_1     | [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xa71f83]
concourse-web_1     |
concourse-web_1     | goroutine 3132 [running]:
concourse-web_1     | github.com/concourse/atc/exec.ValidatingConfigFetcher.FetchConfig(0x0, 0x0, 0xc422945460, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/task_config_fetcher.go:205 +0x73
concourse-web_1     | github.com/concourse/atc/exec.(*ValidatingConfigFetcher).FetchConfig(0xc42038ccd0, 0xc422945460, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	<autogenerated>:1 +0xb3
concourse-web_1     | github.com/concourse/atc/exec.DeprecationConfigFetcher.FetchConfig(0x10c5cda0, 0xc42038ccd0, 0x10c4fe60, 0xc420142c30, 0xc422945460, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/task_config_fetcher.go:87 +0x92
concourse-web_1     | github.com/concourse/atc/exec.(*DeprecationConfigFetcher).FetchConfig(0xc422945360, 0xc422945460, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	<autogenerated>:1 +0xc5
concourse-web_1     | github.com/concourse/atc/exec.(*FetchConfigAction).Run(0xc4248eeb60, 0x10c66620, 0xc4206233e0, 0xc422945460, 0xc4202769c0, 0xc42296c360, 0xc4245185a0, 0xc42002d180)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/fetch_config_action.go:36 +0x76
concourse-web_1     | github.com/concourse/atc/exec.(*ActionsStep).Run(0xc4245185a0, 0xc4202769c0, 0xc42296c360, 0xf3a3c0, 0xc424518550)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/actions_step.go:61 +0xd5
concourse-web_1     | github.com/concourse/atc/exec.(*OnSuccessStep).Run(0xc424518550, 0xc4202769c0, 0xc42296c360, 0xc424518550, 0x0)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/on_success.go:44 +0x4c
concourse-web_1     | github.com/concourse/atc/exec.(*OnSuccessStep).Run(0xc4220ac0f0, 0xc4202769c0, 0xc42296c180, 0xc4220ac0f0, 0x0)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/on_success.go:57 +0x102
concourse-web_1     | github.com/concourse/atc/exec.(*OnSuccessStep).Run(0xc420143180, 0xc4202769c0, 0xc420276a20, 0x10c4de60, 0xc4225082a0)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/on_success.go:57 +0x102
concourse-web_1     | github.com/tedsuo/ifrit.(*process).run(0xc420562e80)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/tedsuo/ifrit/process.go:71 +0x49
concourse-web_1     | created by github.com/tedsuo/ifrit.Background
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/tedsuo/ifrit/process.go:49 +0x117
concourse-worker_1  | {"timestamp":"1508798718.357427120","source":"worker","message":"worker.beacon.restarting","log_level":2,"data":{"error":"wait: remote command exited without exit status or exit signal","session":"4"}}
```

Expected result:
That it gives me an error that either the file or config parameter is missing.

* Concourse version: Occurs in 3.3.2 up to and including 3.5.0
* Deployment type (BOSH/Docker/binary): Docker
* Infrastructure/IaaS: Docker-compose
* Browser (if applicable): -
* Did this used to work? No

**Comment by JohannesRudolph** [comment]
I am experiencing a similar issue on Concourse 3.6 deployed on Docker: 

```
anic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xa784c3]

goroutine 5016 [running]:
github.com/concourse/atc/exec.ValidatingConfigFetcher.FetchConfig(0x0, 0x0, 0xc4216fb080, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/task_config_fetcher.go:200 +0x73
github.com/concourse/atc/exec.(*ValidatingConfigFetcher).FetchConfig(0xc4216f5530, 0xc4216fb080, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	<autogenerated>:1 +0xb3
github.com/concourse/atc/exec.DeprecationConfigFetcher.FetchConfig(0x1177df60, 0xc4216f5530, 0x11770ea0, 0xc421658ae0, 0xc4216fb080, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/task_config_fetcher.go:82 +0x92
github.com/concourse/atc/exec.(*DeprecationConfigFetcher).FetchConfig(0xc4216fae60, 0xc4216fb080, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	<autogenerated>:1 +0xc5
github.com/concourse/atc/exec.(*FetchConfigAction).Run(0xc4203e3d40, 0x117877e0, 0xc421658a80, 0xc4216fb080, 0xc423b8f800, 0xc423b8f860, 0xc424b6f708, 0xc4240931f8)
	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/fetch_config_action.go:36 +0x76
github.com/concourse/atc/exec.(*ActionsStep).Run(0xc42171c4b0, 0xc423b8f800, 0xc423b8f860, 0x10fd58d8, 0x0)
	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/actions_step.go:61 +0xd5
github.com/tedsuo/ifrit.(*process).run(0xc421924400)
	/tmp/build/9674af12/concourse/src/github.com/tedsuo/ifrit/process.go:71 +0x49
created by github.com/tedsuo/ifrit.Background
	/tmp/build/9674af12/concourse/src/github.com/tedsuo/ifrit/process.go:49 +0x117
ubuntu@jumphost:~/infrastructure/concourse$ 
```

I'm not sure if I can support your conclusion that it's caused by a "rogue" task config but I'll investigate.

**Comment by JohannesRudolph** [comment]
Fortunately I was able to get my ATC back to life by manually deleting the pipeline from the atc database using `psql`...

I can confirm that also in my case the issue was a misplaced `file` property of a task (wrong indent in an aggregate block). I think the desirable behaviour here is: 
- reject pipelines with such an invalid task config
- fix the particular nil pointer issue here
- long shot: mark a pipeline as "failed" or invalid when concourse fails to parse it from the database instead of sending ATC into an infinite crash/retry loop

**Comment by andrew-edgar** [comment]
Any update on this?  seems very bad that a poorly configured pipeline can bring down your whole concourse env.

We just hit this same issue.

**Comment by DanielJonesEB** [comment]
Just been able to reproduce this on 3.9.0 running in Docker Compose, through our own ineptitude.

Basically just `set-pipeline` with a `task` with an empty `config` block:

```yaml
...
- name: compile
  plan:
  - get: golang-image
    trigger: true
    passed:
    - build-image
  - task: lint
    image: golang-image
    config:
```

Results in:
```
concourse-web_1     | {"timestamp":"1519143463.977886200","source":"atc","message":"atc.build-tracker.track.do.get.finished","log_level":1,"data":{"build":2,"job":"compile","name":"golang-image","pipeline":"training","session":"41.1.3.5","version-info":{"Version":{"digest":"sha256:43bf260a67f24f0b1eafddfdf5637903dccc5fe69a8e14e1c080de409af32a6b"},"Metadata":[{"name":"image","value":"sha256:41dea"}]}}}
concourse-web_1     | panic: runtime error: invalid memory address or nil pointer dereference
concourse-web_1     | [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xa69c53]
concourse-web_1     |
concourse-web_1     | goroutine 259 [running]:
concourse-web_1     | github.com/concourse/atc/exec.ValidatingConfigFetcher.FetchConfig(0x0, 0x0, 0xc421fb03e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/task_config_fetcher.go:200 +0x73
concourse-web_1     | github.com/concourse/atc/exec.(*ValidatingConfigFetcher).FetchConfig(0xc420762a80, 0xc421fb03e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	<autogenerated>:1 +0xb3
concourse-web_1     | github.com/concourse/atc/exec.DeprecationConfigFetcher.FetchConfig(0x13166400, 0xc420762a80, 0x13151140, 0xc4206b34a0, 0xc421fb03e0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/task_config_fetcher.go:82 +0x92
concourse-web_1     | github.com/concourse/atc/exec.(*DeprecationConfigFetcher).FetchConfig(0xc421fb03a0, 0xc421fb03e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
concourse-web_1     | 	<autogenerated>:1 +0xc5
concourse-web_1     | github.com/concourse/atc/exec.(*FetchConfigAction).Run(0xc421fae270, 0x1317d640, 0xc4206b3440, 0xc421fb03e0, 0xc4205ede00, 0xc4225d27e0, 0xc422468f50, 0xc420300400)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/fetch_config_action.go:36 +0x76
concourse-web_1     | github.com/concourse/atc/exec.(*ActionsStep).Run(0xc422468f50, 0xc4205ede00, 0xc4225d27e0, 0x19ad520, 0xc422468f00)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/actions_step.go:61 +0xd5
concourse-web_1     | github.com/concourse/atc/exec.(*OnSuccessStep).Run(0xc422468f00, 0xc4205ede00, 0xc4225d27e0, 0xc422468f00, 0x0)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/on_success.go:44 +0x4c
concourse-web_1     | github.com/concourse/atc/exec.(*OnSuccessStep).Run(0xc422468460, 0xc4205ede00, 0xc422241ce0, 0xc422468460, 0x0)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/on_success.go:58 +0x102
concourse-web_1     | github.com/concourse/atc/exec.(*OnSuccessStep).Run(0xc420145310, 0xc4205ede00, 0xc4205ede60, 0xc4203e3f80, 0x13175b40)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/concourse/atc/exec/on_success.go:58 +0x102
concourse-web_1     | github.com/tedsuo/ifrit.(*process).run(0xc4205eba00)
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/tedsuo/ifrit/process.go:71 +0x49
concourse-web_1     | created by github.com/tedsuo/ifrit.Background
concourse-web_1     | 	/tmp/build/9674af12/concourse/src/github.com/tedsuo/ifrit/process.go:49 +0x117
```

**Comment by DanielJonesEB** [comment]
Reproducible environment in case you need it: https://github.com/EngineerBetter/b0rkline

**Comment by mhuangpivotal** [comment]
We fixed the validation so that empty config or file would result in validation errors. Previously, image artifact was considered as an alternative to task config.

**Comment by staylor14** [comment]
Is there a way y'all could cut 3.9.1 without this fix, and put this in a 3.9.2?  We really need the optional input feature in 3.9 on the CF Release Integration team.

**Comment by DanielJonesEB** [comment]
@staylor14 Airplane.ci isn't affected, but some of our friends who run a more multi-tenant commercial hosted Concourse offering are at quite a risk, I believe.

**Comment by staylor14** [comment]
@DanielJonesEB I'm not sure I understand how it slows down your work on this bug to cut a 3.9.1 release now.  

If I'm reading this correctly, the bug's been present since at least 3.3.2, so this isn't a regression (AFAICT).  Why not just cut 3.9.1 with what's ready to go, and cut 3.9.2 whenever this bug is fixed?

**Comment by DanielJonesEB** [comment]
Sorry @staylor14, I may not have been following the context of what's going on fully. I meant only to point out that for some commercial interests (who I shalln't name as it's a this-will-take-down-their-product bug) getting this fix is a high priority. I hadn't realised all the other 3.9.1  stuff was ready.

My intention was to provide information that the Concourse team may not be privy to so they could factor that into their release-cutting decision. Not that it's my place to say, but if cutting 3.9.1 _now_ without the fix doesn't delay 3.9.2 coming with the fix, that seems like a win-win situation.

**Comment by jama-pivotal** [comment]
@staylor14 and @DanielJonesEB this bug is actually "Done" and just not "closed" because we haven't had a chance to accept it while babysitting our pipelines to go green for 3.9.1. To be clear this issue isn't "holding up" the release for 3.9.1; Topgun tests were failing and that's what held us up.

**Comment by vito** [comment]
Woops, this should be closed; 3.9.1 is out now.
